test/mitmproxy/test_proxy.py

```diff
@@ -80,6 +82,29 @@ def test_simple(self):
 
 class TestConnectionHandler:
 
+    def check_proxy_for_mode(self, mode):
+        opts = options.Options(mode=mode)
+        pconf = config.ProxyConfig(opts)
+
+        channel = mock.Mock()
+
+        c = ConnectionHandler(
+            mock.MagicMock(),
+            ("127.0.0.1", 8080),
+            pconf,
+            channel
+        )
+        try:
+            c.handle()
+        except exceptions.ProtocolException as e:
+            pytest.fail(repr(e))
+
+        c = pathoc.Pathoc(("127.0.0.1", 8080))
+        try:
+            c.connect()
+        except exceptions.TcpException:
+            pass
```

Comment:
...speaking of which: What does actually happen now? We should have an assertion here that a log statement is generated or something like that :)

Reply:
The traceback appears, when we run mitmproxy in reverse/upstream mode, setting mitmproxy's address as upstream server's address and then making a request using **pathoc**.

First of all, I create `ConnectionHandler` object, according to this: https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/proxy/server.py#L62-L68
Second of all, I make a request using pathoc to reproduce the issue, we are talking about. 

The thing is two exceptions are raised: `mitmproxy.exceptions.ProtocolException` from mitmproxy side and `mitmproxy.exceptions.TcpException` from pathoc side. I decided to check our code only for `ProtocolException`, so I wrote: _if ProtocolException is raised (it hasn't to be raised), test must be failed_. Since pathoc exception not a business of this issue, I just skip it. That's it.